File: mt5_executor.py
import MetaTrader5 as mt5
import sys
import json
import argparse
from datetime import datetime

def initialize_mt5():
    if not mt5.initialize():
        print(json.dumps({
            "success": False,
            "error": {
                "code": mt5.last_error(),
                "message": "MT5 initialization failed"
            }
        }))
        mt5.shutdown()
        sys.exit(1)

def check_connection():
    try:
        initialize_mt5()
        account_info = mt5.account_info()
        connected = account_info is not None
        
        result = {
            "success": connected,
            "account_info": {
                "login": account_info.login,
                "server": account_info.server,
                "balance": account_info.balance,
                "equity": account_info.equity
            } if connected else None
        }
        
        print(json.dumps(result))
        mt5.shutdown()
        return connected
        
    except Exception as e:
        print(json.dumps({
            "success": False,
            "error": {
                "message": str(e),
                "type": "CONNECTION_ERROR"
            }
        }))
        return False

def execute_trade(signal_data):
    try:
        initialize_mt5()
        
        # Validate required fields
        required_fields = ['symbol', 'volume', 'action', 'tp', 'sl']
        for field in required_fields:
            if field not in signal_data:
                raise ValueError(f"Missing required field: {field}")
        
        symbol = signal_data['symbol']
        volume = float(signal_data['volume'])
        action = signal_data['action']
        tp = float(signal_data['tp'])
        sl = float(signal_data['sl'])
        
        # Validate symbol
        symbol_info = mt5.symbol_info(symbol)
        if symbol_info is None:
            raise ValueError(f"Symbol {symbol} not found")
            
        if not symbol_info.visible:
            if not mt5.symbol_select(symbol, True):
                raise ValueError(f"Symbol {symbol} selection failed")
        
        # Get current price
        tick = mt5.symbol_info_tick(symbol)
        if tick is None:
            raise ValueError(f"Failed to get price for {symbol}")
        
        # Determine order type and price
        order_type = mt5.ORDER_TYPE_BUY if action == 'BUY' else mt5.ORDER_TYPE_SELL
        price = tick.ask if action == 'BUY' else tick.bid
        
        # Prepare trade request
        request = {
            "action": mt5.TRADE_ACTION_DEAL,
            "symbol": symbol,
            "volume": volume,
            "type": order_type,
            "price": price,
            "sl": sl,
            "tp": tp,
            "deviation": 10,
            "magic": 202504,
            "comment": "Teaka AutoExec",
            "type_time": mt5.ORDER_TIME_GTC,
            "type_filling": mt5.ORDER_FILLING_IOC,
        }
        
        # Send order
        result = mt5.order_send(request)
        
        if result.retcode != mt5.TRADE_RETCODE_DONE:
            raise Exception(f"Order failed: {result.comment}")
        
        response = {
            "success": True,
            "order": {
                "ticket": result.order,
                "volume": result.volume,
                "price": result.price,
                "comment": result.comment
            }
        }
        
        print(json.dumps(response))
        mt5.shutdown()
        
    except ValueError as e:
        print(json.dumps({
            "success": False,
            "error": {
                "type": "VALIDATION_ERROR",
                "message": str(e)
            }
        }))
        mt5.shutdown()
        
    except Exception as e:
        print(json.dumps({
            "success": False,
            "error": {
                "type": "EXECUTION_ERROR",
                "message": str(e)
            }
        }))
        mt5.shutdown()

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('--check-connection', action='store_true')
    args = parser.parse_args()
    
    if args.check_connection:
        check_connection()
    else:
        try:
            # Read signal data from stdin
            signal_data = json.loads(sys.stdin.read())
            execute_trade(signal_data)
        except json.JSONDecodeError:
            print(json.dumps({
                "success": False,
                "error": {
                    "type": "INVALID_JSON",
                    "message": "Failed to parse signal data"
                }
            }))

File: mt5Executor.js
import { PythonShell } from 'python-shell';
import { logger } from '../utils/logger.js';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export class MT5Executor {
  constructor() {
    this.pythonPath = process.env.PYTHON_PATH || 'python';
    this.scriptPath = path.join(__dirname, '..', 'python', 'mt5_executor.py');
  }

  async executeSignal(signal) {
    try {
      logger.info('Executing MT5 signal', { signal });

      const options = {
        mode: 'json',
        pythonPath: this.pythonPath,
        pythonOptions: ['-u'], // unbuffered output
        scriptPath: path.dirname(this.scriptPath),
        args: [],
      };

      const executionData = {
        symbol: signal.symbol,
        volume: signal.volume || 0.1,
        action: signal.type,
        tp: signal.takeProfit.toNumber(),
        sl: signal.stopLoss.toNumber(),
      };

      const pyshell = new PythonShell('mt5_executor.py', options);

      return new Promise((resolve, reject) => {
        // Send signal data to Python script
        pyshell.send(JSON.stringify(executionData));

        pyshell.on('message', (message) => {
          logger.info('MT5 execution result:', message);
          resolve(message);
        });

        pyshell.on('error', (error) => {
          logger.error('MT5 execution error:', error);
          reject(error);
        });

        pyshell.end((err, code, signal) => {
          if (err) {
            logger.error('MT5 script error:', err);
            reject(err);
          }
        });
      });
    } catch (error) {
      logger.error('MT5 execution failed:', error);
      throw error;
    }
  }

  async checkConnection() {
    try {
      const options = {
        mode: 'json',
        pythonPath: this.pythonPath,
        pythonOptions: ['-u'],
        scriptPath: path.dirname(this.scriptPath),
        args: ['--check-connection'],
      };

      const result = await new Promise((resolve, reject) => {
        PythonShell.run('mt5_executor.py', options, (err, output) => {
          if (err) reject(err);
          resolve(output);
        });
      });

      return result[0]?.connected || false;
    } catch (error) {
      logger.error('MT5 connection check failed:', error);
      return false;
    }
  }
}